Login/authentication enhancements

Wauplin opened the pull request:
This PR do various things related to authentication. I started it to support Google Colab secrets and ended up rewriting some docs and deprecating some legacy code.

**List of changes:**
1. New authentication section in the quick-start guide. Especially:
    - document how to login via CLI
    - encourage login with env variable for Google Colab and Spaces
    - discourage hardcoded tokens
    - warnings about token permission and leaks
    - => this section is in the quick-start guide. I hesitated to have a dedicated guide/page for it but I don't think it's worth it (not much content + we would still have some "authentication" content in the quick start so let's keep everything together)
2. **Automatically read `HF_TOKEN` secret when running in a colab notebook**! :fire: 
3. `logout` now raises an error if token is deleted but user was logged in via environment variable / secret value. I think this is important otherwise a user can think they are completely logged out of a machine when it's not the case.
4. Started to deprecate `HfFolder` in favor of `get_token`, `login` and `logout`.
    - `huggingface_hub.get_token` feels much more intuitive `huggingface_hub.Hffolder.get_token`
    - `login` in favor of `HfFolder.save_token` => saves token but also checks token validity. I don't think an external lib should save the token without checks.
    - `logout` in favor of `HfFolder.delete_token`  => deletes token but also checks if the user is really logged out (i.e. no env variable)
    - => for now, `HfFolder` is not officially deprecated. I will first release this, then adapt `transformers`/`datasets`/`gradio` codebase and then deprecate it officially in the next release (and most likely remove in `v1.0`).
5. ignore token if empty string (`""`) and consider as None
6. adapt tests

**TODO:**
- [ ] check documentation in other languages
- [ ] fix remaining tests

![image](https://github.com/huggingface/huggingface_hub/assets/11801849/60206997-64c1-4e8f-9127-a343f091cd68)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1895). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

LysandreJik commented on docs/source/en/quick-start.md:
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
-Once you are logged in, all requests to the Hub -even methods that don't necessarily require authentication- will use your
-access token by default. If you want to disable implicit use of your token, you should set the
-`HF_HUB_DISABLE_IMPLICIT_TOKEN` environment variable.
+Please be careful when passing tokens as parameter. It is always best practice to load the token from a secure vault instead of hardcoding it in your codebase or notebook. Hardcoded tokens present a major risk of leak if you share your code inadvertently.
```
Yes! We'll probably need to do a few updates in the libraries that depend on it as we were encouraging this for now

LysandreJik reviewed (approved):
Yes, agree with the refactor! Nice to see the updates. I agree that moving away from `token=xxx` as a default within code is better for security; we'll need to update downstream documentation as well as I'm pretty sure we have a lot of references to `token` in `transformers`/`diffusers`.

LGTM!

stevhliu reviewed (approved):
LGTM!

I did a quick search of the Diffusers/Transformers repos with `repo:huggingface/transformers token= language:Markdown` and the only docs I found that reference `token=...` is the NLLB doc [here](https://huggingface.co/docs/transformers/v4.35.2/en/model_doc/nllb#generating-from-any-other-language-than-english).

julien-c commented:
> I agree that moving away from `token=xxx` as a default within code is better for security

Honestly, let's not go overboard with this, given anyways both on the Hub and on GH, any uploaded HF token will soon be auto-invalidated

So honestly i'm not sure it's a huge issue

Wauplin commented on docs/source/en/quick-start.md:
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
-Once you have your User Access Token, run the following command in your terminal:
+<Tip>
+
+Tokens can have `read` or `write` permissions. Make sure to have a `write` access token if you want to create or edit a repository. Otherwise, it's best to generate a `read` token to reduce risks in case you leak your token inadvertently.
```
Ahah, yes that sounds better :smile:

Wauplin commented on docs/source/en/quick-start.md:
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
+
+</Tip>
+
+Authentication via an environment variable or a secret has priority over the token stored on your machine.
```
What I meant is if a token is stored on the machine (via `huggingface-cli login` AND the environment variable is set, then the environment variable is used (so "it has priority").

Wauplin commented:
Thanks both for the reviews! I addressed all of the comments I think. 
I have updated the Google Colab integration to handle errors nicely. The logic is:
- if user has defined a `HF_TOKEN` secret and granted access to the notebook => use it => normal use case
- if user has defined a `HF_TOKEN` secret but did not grant access to the notebook, a pop-up will ask permission
   - if permission is granted by the user => use the token. Next calls will not need to re-request the access.
   - if permission is NOT granted =>  warn the user they'll not be authenticated + set flag to ensure we do not ask again + do not fail (returns `None` as token)
- if `HF_TOKEN` is not defined in the user secrets => warn to say "hey it's recommended to define a token" + return `None` 
- if anything else happen => warn "You are not authenticated and we don't know why. Please report on GitHub" + return `None`

Wauplin commented:
@LysandreJik could you re-review logic for `_get_token_from_google_colab` in `src/huggingface_hub/_login.py` please? (see https://github.com/huggingface/huggingface_hub/pull/1895#issuecomment-1850356442) It's the only piece of real logic that I have updated. Apart from that, tests look ok and we should be good to merge.

LysandreJik reviewed (approved):
Impressive PR @Wauplin! This looks good to me. Thanks for working on this.

Wauplin commented:
Thanks for the last review @LysandreJik, let's merge this! :)
